src/python/pants/core/goals/publish.py

```diff
@@ -0,0 +1,160 @@
+
+    def __init__(self, published_packages: Iterable[PublishedPackage]):
+        self.publishes = FrozenOrderedSet(published_packages)
+
+
+def _can_package(target: Target, union_membership: UnionMembership):
+    package_request_types = cast(Iterable[Type[PackageFieldSet]], union_membership[PackageFieldSet])
+
+    for fieldset in package_request_types:
+        if fieldset.is_applicable(target):
+            return fieldset
+
+    return False
+
+
+@goal_rule
+async def publish(
+    targets: Targets,
+    union_membership: UnionMembership,
+) -> Publish:
+    publishable_targets: List[Tuple[Target, PackageFieldSet]] = []
+
+    # Retrieve publishable targets and their package type.
+    for target in targets:
+        if PublishTargetFieldSet.is_applicable(target):
+            package_fieldset = _can_package(target, union_membership)
+            if package_fieldset:
+                publishable_targets.append((target, package_fieldset))
+            else:
+                logger.warn(
+                    f"Unable to publish {target.address} as it is not a packageable target."
+                )
+                return Publish(exit_code=1)
+
+    # Get the publish targets (destination)
+    publish_targets_set = cast(
+        Iterable[Iterable[PublishTarget]],
+        await MultiGet(
+            Get(
+                Targets,
+                UnparsedAddressInputs,
+                publishable[PublishTargetField].to_unparsed_address_inputs(),
+            )
+            for (publishable, _) in publishable_targets
+        ),
+    )
+
+    # get the built packages.
+    built_packages = await MultiGet(
+        Get(BuiltPackage, PackageFieldSet, package_fieldset.create(target))
+        for (target, package_fieldset) in publishable_targets
+    )
+
+    # Build a list of requests, ensuring that the publishable is valid for the
+    # publish target. Error if not.
+    requests: List[PublishRequest] = []
+
+    for (target, _), publish_targets, built_package in zip(
+        publishable_targets, publish_targets_set, built_packages
+    ):
+        for publish_target in publish_targets:
+            if not publish_target.publishee_fieldset_type.is_applicable(target):
+                logger.warn(
+                    f"Cannot publish {target.address} to {publish_target.address}."
+                )
+                return Publish(exit_code=1)
+            fieldset = publish_target.publishee_fieldset_type.create(target)
+
+            logger.info(
+                f"Publishing {target.address} to {publish_target.address}"
+            )
+            requests.append(
+                publish_target.publish_request_type(
+                    built_package,
+                    fieldset,
+                    publish_target,
+                )
+            )
+
+    packages = await MultiGet(
+        Get(PublishedPackage, PublishRequest, request) for request in requests
+    )
+
+    # ???
+    PublishedPackageSet(packages)
```

Comment:
Not quite sure what you're meant to do with the resulting values from the concrete implementations?

Reply:
Only for logging purposes in the publish goal, I'd say.

Reply:
Yea, this is likely a good place to render something to the `Console` that lists the things that were published.

Reply:
Better is logging imo:

https://github.com/pantsbuild/pants/blob/fd71177d18bd025e87e14322adf8a5abb8a02061/src/python/pants/core/goals/package.py#L102-L109

We've been using logging when it describes a side-effect like this.

Reply:
By this point the publish will already have happened in the background, so that's the relevant side-effect logging (see other comment). So that's why this felt like a good place for a summary, similar to: https://github.com/pantsbuild/pants/blob/e1dd9812fcabf429549c22c2ef8c6b71aacab2c1/src/python/pants/core/goals/lint.py#L260-L275